### PR TITLE
[PACT-650] Added support for creators.

### PR DESCRIPTION
**Pull request by tna-erasmos**
Also reduced the size of many scenarios by using a base expected record.

### CODE REVIEW

**Comment by tna-erasmos** [comment]
Please note that my implementation follows my comment [here](https://national-archives.atlassian.net/browse/PACT-650?focusedCommentId=37706), **not** the description in the story; the former is more detailed and matches the revised prototype.

As usual, you can watch [the demo video](https://drive.google.com/drive/folders/1d2WcgoKUC3q6cl7_O5bcYc5-ST4-UTot), which represents my QA as well.

File: app/uk/gov/nationalarchives/omega/editorial/support/DisplayedCreator.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.gov.nationalarchives.omega.editorial.support
+
+import uk.gov.nationalarchives.omega.editorial.models.Creator
+import uk.gov.nationalarchives.omega.editorial.models.Creator.{ CreatorTypeCorporateBody, CreatorTypePerson }
+
+case class DisplayedCreator(id: String, name: String)
```
Comment (felixgb) [inline app/uk/gov/nationalarchives/omega/editorial/support/DisplayedCreator.scala:27]: Do need this case class, or could `fromCreator` be a method on  `Creator`? Possibly called `formattedName` or some such?

Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/support/DisplayedCreator.scala:27]: I'm afraid I don't understand; could you sketch it out?

Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/support/DisplayedCreator.scala:27]: Now, I don't think we'd ever need to convert a `DisplayedCreator` into a `Creator`; it's just for presentation.

File: app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.gov.nationalarchives.omega.editorial.services
+
+import com.google.inject.ImplementedBy
+import uk.gov.nationalarchives.omega.editorial.models._
+
+@ImplementedBy(classOf[StubReferenceDataService])
+trait ReferenceDataService {
```
Comment (felixgb) [inline app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala:28]: Do we need this trait with a default implementation?

I'm aware that we override this trait in the tests, however I think we could equally make test reference data class extend `StubReferenceDataService` - as long we don't do side effecting work in the constructor (which I would flag on review anyway).

I think that creating a trait where the only non-default implementation is in the tests seems a bit like altering the application code for the test's sake. It's also a pain to have to change the abstract method implementation every time a function signature changes.

File: app/uk/gov/nationalarchives/omega/editorial/support/DisplayedCreator.scala (hunk 1 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/support/DisplayedCreator.scala:27]: I understand it now; let me look into it, @felixgb.

File: app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala (hunk 2 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala:28]: I'll look into it. But it's a pattern I've seen many times elsewhere.

File: app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala [hunk 3]
```diff
@@ -45,18 +45,19 @@ case class EditSetRecord(
   background: String,
   custodialHistory: String,
   relatedMaterial: Seq[RelatedMaterial] = Seq.empty,
-  separatedMaterial: Seq[SeparatedMaterial] = Seq.empty
+  separatedMaterial: Seq[SeparatedMaterial] = Seq.empty,
+  creatorIDs: Seq[String] = Seq.empty
```
Comment (rwalpole) [inline app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala:49]: The `EditSetRecord` should have a sequence of "creators" not "creatorIDs".

File: app/uk/gov/nationalarchives/omega/editorial/support/DisplayedCreator.scala (hunk 1 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/support/DisplayedCreator.scala:27]: Yes, good idea - `DisplayedCreator` is unnecessary.

Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/support/DisplayedCreator.scala]: Changed.

File: app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala (hunk 2 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala:28]: I tried it - you're right; the trait isn't needed.

Will change.

Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala]: Changed.

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala [hunk 4]
```diff
@@ -321,14 +296,47 @@ class EditSetController @Inject() (
...
   }
 
+  private def prepareForDisplay(originalEditSetRecord: EditSetRecord): EditSetRecord =
+    Seq[RecordTransformer](prepareCreatorIDs, preparePlaceOfDeposit).foldLeft(originalEditSetRecord)(
+      (editSetRecord, transformer) => transformer(editSetRecord)
+    )
+
+  private def prepareCreatorIDs(editSetRecord: EditSetRecord): EditSetRecord = {
+    val recognisedCreatorIds = editSetRecord.creatorIDs.filter(isCreatorRecognised)
+    val updatedCreatorIds = if (recognisedCreatorIds.nonEmpty) recognisedCreatorIds else List("")
```
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala:279]: Good catch: an empty list will do.

Will change.

File: app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala (hunk 3 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala:49]: 👀

Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala:49]: I would have thought that the creator IDs were just pointers to reference data, similar to `legalStatus` and `placeOfDeposit`; the reference data in the case is the combined list of corporate bodies and persons.

Even if these came over the wire, the application would ignore all fields except the `id`.

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala]: Changed.

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala [hunk 5]
```diff
@@ -296,14 +269,47 @@ class EditSetController @Inject() (
...
   }
 
+  private def prepareForDisplay(originalEditSetRecord: EditSetRecord): EditSetRecord =
+    Seq[RecordTransformer](prepareCreatorIDs, preparePlaceOfDeposit).foldLeft(originalEditSetRecord)(
+      (editSetRecord, transformer) => transformer(editSetRecord)
+    )
+
+  private def prepareCreatorIDs(editSetRecord: EditSetRecord): EditSetRecord = {
+    val recognisedCreatorIds = editSetRecord.creatorIDs.filter(isCreatorRecognised)
+    val updatedCreatorIds = if (recognisedCreatorIds.nonEmpty) recognisedCreatorIds else List.empty
```
Comment (felixgb) [inline app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala:279]: Does this expression do anything now? Won't it always evaluate to `recognisedCreatorIds` or the equivalent `List.empty`?

Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala:279]: You're correct. I'll change.

Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala]: Changed.

File: app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala (hunk 3 above)
Comment (rwalpole) [reply app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala:49]: If we want to make that distinction @tna-erasmos then we should rename `legalStatus` and `placeOfDeposit` to be consistent.

Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala:49]: So, in `EditSetRecord`, you'd like for me to rename `legalStatus` to `legalStatusID` and `placeOfDeposit` to `placeOfDepositID`?

Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala:49]: I'll assume so and make the change.

Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala:49]: Changed, @rwalpole.

**Review by felixgb — APPROVED** [review]

File: app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala (hunk 3 above)
Comment (rwalpole) [reply app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala:49]: Thanks @tna-erasmos

**Pull request merged**